File: sdk/python/src/ovara_sdk/canon.py
"""Canonical signed-payload encoding — byte-for-byte identical to the
gateway's internal/identity/canon.go lpBuilder.

Format:
    lp(s)    = u32be(len(utf8(s))) || utf8(s)   — strings, always present
    i64(v)   = u64be(v)                          — unix timestamps, ints
    lparr(a) = u32be(count) || lp(item)*         — string arrays

A signature over these bytes identifies exactly one semantic object:
field boundaries, array boundaries, and integers are unambiguous — no
pipe-join or %v formatting quirks.
"""

from __future__ import annotations

import struct
from typing import Iterable, Optional


def _lp(s: str) -> bytes:
    b = s.encode("utf-8")
    return struct.pack(">I", len(b)) + b


def _lparr(items: Iterable[str]) -> bytes:
    out = struct.pack(">I", 0)
    count = 0
    body = b""
    for s in items:
        body += _lp(s)
        count += 1
    return struct.pack(">I", count) + body


def _i64(v: int) -> bytes:
    return struct.pack(">Q", v & 0xFFFFFFFFFFFFFFFF)


def hop_payload(
    issuer: str,
    subject_id: str,
    audience: str,
    resource_scope: str,
    actions: Optional[Iterable[str]],
    expires_at_unix: int,
    delegated_at_unix: int,
    nonce: str,
    prev_sig_hex: str,
) -> bytes:
    """Canonical delegation-hop payload (matches validator.go hopPayload)."""
    return (
        _lp(issuer)
        + _lp(subject_id)
        + _lp(audience)
        + _lp(resource_scope)
        + _lparr(actions or [])
        + _i64(expires_at_unix)
        + _i64(delegated_at_unix)
        + _lp(nonce)
        + _lp(prev_sig_hex)
    )


def lease_payload(
    lease_id: str,
    issuer: str,
    subject: str,
    audience: str,
    allowed_actions: Optional[Iterable[str]],
    resource_scope: str,
    expiry_unix: int,
    issued_at_unix: int,
    delegation_depth: int,
) -> bytes:
    """Canonical capability-lease payload (matches validator.go leasePayload)."""
    return (
        _lp(lease_id)
        + _lp(issuer)
        + _lp(subject)
        + _lp(audience)
        + _lparr(allowed_actions or [])
        + _lp(resource_scope)
        + _i64(expiry_unix)
        + _i64(issued_at_unix)
        + struct.pack(">I", delegation_depth & 0xFFFFFFFF)
    )

File: runtime/gateway/internal/identity/canon.go
package identity

// Canonical signed-payload encoding for leases and delegation hops.
//
// Format: a sequence of fields, each length-prefixed —
//
//	lp(s)  = u32be(len(utf8(s))) || utf8(s)      — strings, always present
//	lpu64  = u64be(v)                            — unix timestamps, ints
//	lparr  = u32be(count) || lp(item)*           — string arrays
//
// Properties that the old pipe/%v format lacked:
//   - field boundaries unambiguous ("x|y" in one field can't collide
//     with ("x","y") across two)
//   - arrays unambiguous (["a b"] != ["a","b"] — different byte counts)
//   - deterministic across languages: byte counts and big-endian ints,
//     no formatting quirks (Go %v, JSON escaping) to replicate.
//
// Every security-relevant field is inside the signed bytes; a signature
// identifies exactly one semantic object.
//
// The same canonical format is implemented by the Python SDK
// (ovara_sdk.canon) — cross-language vectors live in
// delegation_vectors_test.go / sdk/python.

import "encoding/binary"

// lpBuilder accumulates a length-prefixed canonical payload.
type lpBuilder struct {
	buf []byte
}

func (b *lpBuilder) str(s string) *lpBuilder {
	var l [4]byte
	binary.BigEndian.PutUint32(l[:], uint32(len(s)))
	b.buf = append(b.buf, l[:]...)
	b.buf = append(b.buf, s...)
	return b
}

func (b *lpBuilder) strs(ss []string) *lpBuilder {
	var l [4]byte
	binary.BigEndian.PutUint32(l[:], uint32(len(ss)))
	b.buf = append(b.buf, l[:]...)
	for _, s := range ss {
		b.str(s)
	}
	return b
}

func (b *lpBuilder) i64(v int64) *lpBuilder {
	var l [8]byte
	binary.BigEndian.PutUint64(l[:], uint64(v))
	b.buf = append(b.buf, l[:]...)
	return b
}

func (b *lpBuilder) u32(v uint32) *lpBuilder {
	var l [4]byte
	binary.BigEndian.PutUint32(l[:], v)
	b.buf = append(b.buf, l[:]...)
	return b
}
